### PR TITLE
ORPO: Avoid catastrophic cancellation in loss function

### PR DESCRIPTION
**Pull request by hartmans**
I was trying to use ORPO to train models where the differences between the chosen and rejected samples was very small.  The typical use case was trying to correct the behavior of a model instructed to proofread or edit long passages of text. Examples of preference improvements I was trying to correct included avoiding cases where the model skipped forward in the text.

My training runs were blowing up; typically I would get one loss of NAN, then everything would go to +INF or -INF, and of course at that point it was all useless.

I dug into the code and discovered that as the logits became large,  the loss function would blow up.

This fix makes my training much more stable.

I made significant use of ChatGPT in my investigation, but reached a point where I independently understood and verified the following:
* I understand how as the logits increase, we can run into catastrophic cancellation with the original code.
* I am confident that the formulation of the computation I introduced is correct given infinite precision
* I understand how the computation I introduced fixes the specific stability problem I was running into

I have tested the change, but I don't have a good feel for how to reduce my datasets into a form that I would be willing to share and that could be run reasonably in your tests.

# What does this PR do?

Use a numerically more stable form of log(1-exp(x)) to avoid catastrophic cancellation when the probability of a model selecting the chosen or rejected sample is very high.  Also, compute the logs ratio in float32 to increase the stability.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [N/A ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qgallouedec** [comment]
That's quite smart! I did a quick benchmark, and it looks good to me:

```python
import torch

def before(x):
    return torch.log1p(-torch.exp(x))

def after(x):
    # branch at -ln 2 ~ -0.693 to avoid cancellation
    t = -0.6931471805599453
    return torch.where(x < t, torch.log1p(-torch.exp(x)), torch.log(-torch.expm1(x)))

def log_expm1(x):
    return torch.log(-torch.expm1(x))

# plot to verify
if __name__ == "__main__":
    import matplotlib.pyplot as plt
    import numpy as np

    # Log-spaced domains excluding zero
    x = -torch.logspace(-7.5, 1.5, 1000)
    y1 = before(x)
    y2 = after(x)
    y3 = log_expm1(x)
    print(y1)

    plt.plot(x.numpy(), y1.numpy(), label="before")
    plt.plot(x.numpy(), y2.numpy(), label="after", linestyle="dashed")
    plt.plot(x.numpy(), y3.numpy(), label="log_expm1", linestyle="dotted")
    plt.xscale("symlog", linthresh=1e-8)
    # plt.yscale("symlog", linthresh=1e-20)
    plt.legend()
    plt.grid()
    plt.title("Numerical stability improvement for log(1 - exp(x))")
    plt.xlabel("x")
    plt.ylabel("log(1 - exp(x))")
    plt.savefig("bench.png")
```

`torch.log(-torch.expm1(x))` not stable below -10

pos

`torch.log1p(-torch.exp(x))` not stable above -1e-6

pos2

File: trl/experimental/orpo/orpo_trainer.py [hunk 1]
```diff
@@ -73,6 +73,12 @@
 logger = logging.get_logger(__name__)
 
 
+def log1mexp(x: torch.FloatTensor) -> torch.FloatTensor:
```
Comment (qgallouedec) [inline trl/experimental/orpo/orpo_trainer.py:76]:
```suggestion
def log1mexp(x: torch.FloatTensor) -> torch.FloatTensor:
    """Numerically stable computation of log(1-exp(x))."""
```
